Plutarch/Api/V1/Value.hs

```diff
@@ -161,6 +184,105 @@ instance
   where
   inv a = punsafeCoerce $ PlutusTx.inv (punsafeCoerce a :: Term s (PValue 'Sorted 'NoGuarantees))
 
+-- | Check whether a binary relation holds over 2 sorted 'PValue's.
+pcheckBinRel ::
+  Term
+    s
+    ( (PAsData PInteger :--> PAsData PInteger :--> PBool)
+        :--> PValue 'Sorted any0
+        :--> PValue 'Sorted any1
+        :--> PBool
+    )
+pcheckBinRel = phoistAcyclic $
+  plam $ \f m1 m2 ->
+    let inner = pfix #$ plam $ \self l1 l2 ->
+          pelimList
+            ( \x xs ->
+                pelimList
+                  ( \y ys -> unTermCont $ do
+                      v1dat <- tcont . plet $ psndBuiltin # x
+                      v2dat <- tcont . plet $ psndBuiltin # y
+                      k1 <- tcont . plet $ pfromData $ pfstBuiltin # x
+                      k2 <- tcont . plet $ pfromData $ pfstBuiltin # y
+                      pure $
+                        pif
+                          (k1 #< k2)
+                          (pinnerCheckBinRel # f # pfromData v1dat # AssocMap.pempty #&& self # xs # l2)
+                          $ pif
+                            (k1 #== k2)
+                            ( pinnerCheckBinRel # f # pfromData v1dat # pfromData v2dat #&& self
+                                # xs
+                                # ys
+                            )
+                            $ pinnerCheckBinRel # f # AssocMap.pempty # pfromData v2dat #&& self
+                              # l1
+                              # ys
+                  )
+                  ( pinnerCheckBinRel # f # pfromData (psndBuiltin # x) # AssocMap.pempty
+                      #&& pall
+                        # plam
+                          ( \p ->
+                              let v1dat = psndBuiltin # p
+                               in pinnerCheckBinRel # f # pfromData v1dat # AssocMap.pempty
+                          )
+                        # xs
+                  )
+                  l2
```

Comment:
this is optimized for cpu/mem but not script size - unionWith also uses this whole key ordering logic deep within. I wonder if there's a way to manually do common subexpr elim here for optimality.